### DISCUSSION TITLE
No output with example basic-a2dp-spdif.ino

### DISCUSSION THREAD

**Discussion by sneiss** [pschatzmann/arduino-audio-tools#1493] [Q&A] [answered]
I'm trying to get the example file "basic-a2dp-spdif.ino" to work. 

Hardware: 
ESP32 Wrover in an DevKitC V4

Environment: 
Arduino IDE 2.3.2
Board Manager: esp32 2.0.11
Selected board "ESP32 Dev Module"
Clone of today's main branch of the audio-tools and ESP32-A2DP.
PSRAM disabled, "Default" partition scheme (tried "Huge App" as well)

When compiling the example sketch without any changes, I get a compilation error:
`Compilation error: 'AudioInfo' does not name a type`

I have to swap lines 9 and 10 (include "AudioTools.h" before "BluetoothA2DPSink.h") to compile the file without errors. Uploading works just fine, but I get the follwing error message when running the program (changed to AudioLogger to Debug level to get the full output after I had no sound):
```
[D] AudioSPDIF.h : 133 - bool audio_tools::SPDIFOutput::begin(audio_tools::SPDIFConfig)
[D] AudioSPDIF.h : 251 - void audio_tools::SPDIFOutput::spdif_buf_init()
[D] I2SStream.h : 55 - bool audio_tools::I2SStream::begin(audio_tools::I2SConfig)
[D] AudioStreams.h : 52 - virtual void audio_tools::AudioStream::setAudioInfo(audio_tools::AudioInfo)
[I] AudioTypes.h : 126 - sample_rate: 88200
[I] AudioTypes.h : 127 - channels: 2
[I] AudioTypes.h : 128 - bits_per_sample: 32
[D] I2SESP32.h : 61 - bool audio_tools::I2SDriverESP32::begin(audio_tools::I2SConfigESP32)
[D] I2SESP32.h : 180 - bool audio_tools::I2SDriverESP32::begin(audio_tools::I2SConfigESP32, int, int)
[I] AudioTypes.h : 126 - sample_rate: 88200
[I] AudioTypes.h : 127 - channels: 2
[I] AudioTypes.h : 128 - bits_per_sample: 32
[I] I2SConfigESP32.h : 80 - rx/tx mode: TX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 88200
[I] I2SConfigESP32.h : 84 - bits per sample: 32
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 95 - fixed_mclk: 22579200
[I] I2SConfigESP32.h : 97 - buffer_count:30
[I] I2SConfigESP32.h : 98 - buffer_size:384
[I] I2SConfigESP32.h : 107 - pin_data: 23
[D] I2SESP32.h : 212 - i2s_driver_install
[D] I2SESP32.h : 228 - i2s_set_pin
[D] I2SESP32.h : 239 - i2s_zero_dma_buffer
[D] I2SESP32.h : 243 - begin - started
E (163) I2S: register I2S object to platform failed
```
After connecting a device an starting playback, I get 
```
[D] I2SStream.h : 95 - I2SStream::write: 1536
[D] I2SESP32.h : 96 - size_t audio_tools::I2SDriverESP32::writeBytes(const void*, size_t)
[D] I2SESP32.h : 104 - i2s_write 1536 -> 1536 bytes
...
```
I also tried using PlatformIO to get the example to work. I don't get compilation errors here, but the error message when running the program is the same. Any other example using SPDIF as output works fine. I also tried the "base-generator-a2dp.ino" to check if I had a general problem with A2DP and it works just as expected. 

Is there anything I'm missing in my approach?

**Comment by sneiss** [comment]
I changed the output to analog (44.1kHz, stereo, 16 bit) and I get "something that sounds similar to what I'm playing". Sounds a bit as if there are samples missing. I have similar distortions with the generator-analog example so this might be related to my wiring.

**Comment by pschatzmann** [comment, marked as answer]
With the latest version of A2DP, it is possible to define the output. 
Can you try the following
```C++

#include "AudioTools.h"
#include "BluetoothA2DPSink.h"

AudioInfo info(44100, 2, 16);
SPDIFOutput spdif;
BluetoothA2DPSink a2dp_sink(spdif);

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // setup output
  auto cfg = spdif.defaultConfig();
  cfg.copyFrom(info);
  cfg.buffer_size = 384;
  cfg.buffer_count = 30;
  cfg.pin_data = 23;
  spdif.begin(cfg);

  // Start Bluetooth Audio Receiver
  a2dp_sink.start("a2dp-spdif");

}

void loop() {
  delay(100);
}
```

**Reply by sneiss** [reply]
Just wanted to mark this as answer, thanks a lot! "buffer_size" and "buffer_count" are no members of SPDIFConfig or AudioInfo. So I just used the old way.

**Reply by pschatzmann** [reply]
Good point: I have added the buffer_size and buffer_count in my latest commit, since this is quite a relevant setting, so with the latest release the proposed code should compile as well...

**Reply by muhammadazmi26** [reply]
If you use Spdif as output, bluetooth will not appear when scanned.
But if spdif is replaced with analog output, bluetooth will appear.
Please how is this?

**Reply by sneiss** [reply]
Without any Output from the Logger, this is just guessing. But I suppose your program has an error before reaching the point at which BT is started. Please add some more info. You can also set the AudioLogger to Info or Debug to get more details about what's happening in your system.

**Reply by muhammadazmi26** [reply]
![WhatsApp Image 2024-04-28 at 01 11 22](https://github.com/pschatzmann/arduino-audio-tools/assets/41564644/27235c0b-971a-4092-8fed-6f703e5b2ce4)

![WhatsApp Image 2024-04-28 at 01 19 11](https://github.com/pschatzmann/arduino-audio-tools/assets/41564644/b89692e7-3d70-4449-b19b-ede6bce25d85)


this is code:

/**
 * @file basic-a2dp-audiospdif.ino
 * @brief A2DP Sink with output to SPDIFOutput
 * 
 * @author Phil Schatzmann
 * @copyright GPLv3
 */
#include "AudioTools.h"
#include "BluetoothA2DPSink.h"

AudioInfo info(44100, 2, 16);
SPDIFOutput spdif;
BluetoothA2DPSink a2dp_sink(spdif);

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Debug);

  // setup output
  auto cfg = spdif.defaultConfig();
  cfg.copyFrom(info);
  cfg.buffer_size = 384;
  cfg.buffer_count = 30;
  cfg.pin_data = 23;
  spdif.begin(cfg);

  // Start Bluetooth Audio Receiver
  a2dp_sink.start("a2dp-spdif");

}

void loop() {
  delay(100);
}



================================== Audio Logger:  DEBUG ===================================
ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13260
load:0x40080400,len:3028
entry 0x400805e4
[D] AudioSPDIF.h : 135 - bool audio_tools::SPDIFOutput::begin(audio_tools::SPDIFConfig)
[D] AudioSPDIF.h : 255 - void audio_tools::SPDIFOutput::spdif_buf_init()
[D] I2SStream.h : 62 - bool audio_tools::I2SStream::begin(audio_tools::I2SConfig)
[D] AudioStreams.h : 52 - virtual void audio_tools::AudioStream::setAudioInfo(audio_tools::AudioInfo)
[I] AudioTypes.h : 127 - in: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 - out: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[D] I2SESP32.h : 61 - bool audio_tools::I2SDriverESP32::begin(audio_tools::I2SConfigESP32)
[D] I2SESP32.h : 180 - bool audio_tools::I2SDriverESP32::begin(audio_tools::I2SConfigESP32, int, int)
[I] AudioTypes.h : 127 -  sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] I2SConfigESP32.h : 80 - rx/tx mode: RXTX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 88200
[I] I2SConfigESP32.h : 84 - bits per sample: 32
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 95 - fixed_mclk: 22579200
[I] I2SConfigESP32.h : 97 - buffer_count:30
[I] I2SConfigESP32.h : 98 - buffer_size:384
[I] I2SConfigESP32.h : 107 - pin_data: 23
[D] I2SESP32.h : 212 - i2s_driver_install
[D] I2SESP32.h : 228 - i2s_set_pin
[D] I2SESP32.h : 239 - i2s_zero_dma_buffer
[D] I2SESP32.h : 243 - begin - started
[D] AudioSPDIF.h : 135 - bool audio_tools::SPDIFOutput::begin(audio_tools::SPDIFConfig)
[D] I2SStream.h : 72 - virtual void audio_tools::I2SStream::end()
[D] I2SESP32.h : 86 - void audio_tools::I2SDriverESP32::end()
[D] AudioSPDIF.h : 255 - void audio_tools::SPDIFOutput::spdif_buf_init()
[D] I2SStream.h : 62 - bool audio_tools::I2SStream::begin(audio_tools::I2SConfig)
[D] AudioStreams.h : 52 - virtual void audio_tools::AudioStream::setAudioInfo(audio_tools::AudioInfo)
[I] AudioTypes.h : 127 - out: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[D] I2SESP32.h : 61 - bool audio_tools::I2SDriverESP32::begin(audio_tools::I2SConfigESP32)
[D] I2SESP32.h : 180 - bool audio_tools::I2SDriverESP32::begin(audio_tools::I2SConfigESP32, int, int)
[I] AudioTypes.h : 127 -  sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] I2SConfigESP32.h : 80 - rx/tx mode: RXTX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 88200
[I] I2SConfigESP32.h : 84 - bits per sample: 32
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 95 - fixed_mclk: 22579200
[I] I2SConfigESP32.h : 97 - buffer_count:30
[I] I2SConfigESP32.h : 98 - buffer_size:384
[I] I2SConfigESP32.h : 107 - pin_data: 23
[D] I2SESP32.h : 212 - i2s_driver_install
[D] I2SESP32.h : 228 - i2s_set_pin
[D] I2SESP32.h : 239 - i2s_zero_dma_buffer
[D] I2SESP32.h : 243 - begin - started



================================== Audio Logger:  INFO ===================================
ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13260
load:0x40080400,len:3028
entry 0x400805e4
[I] AudioTypes.h : 127 - in: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 - out: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 -  sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] I2SConfigESP32.h : 80 - rx/tx mode: RXTX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 88200
[I] I2SConfigESP32.h : 84 - bits per sample: 32
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 95 - fixed_mclk: 22579200
[I] I2SConfigESP32.h : 97 - buffer_count:30
[I] I2SConfigESP32.h : 98 - buffer_size:384
[I] I2SConfigESP32.h : 107 - pin_data: 23
[I] AudioTypes.h : 127 - out: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 -  sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] I2SConfigESP32.h : 80 - rx/tx mode: RXTX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 88200
[I] I2SConfigESP32.h : 84 - bits per sample: 32
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 95 - fixed_mclk: 22579200
[I] I2SConfigESP32.h : 97 - buffer_count:30
[I] I2SConfigESP32.h : 98 - buffer_size:384
[I] I2SConfigESP32.h : 107 - pin_data: 23

**Reply by pschatzmann** [reply]
You forgot to activate the ESP32 Logger in the Arduino Tools menu! 
I suggest to set it to Info...

**Reply by muhammadazmi26** [reply]
ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13260
load:0x40080400,len:3028
entry 0x400805e4
[   454][I][esp32-hal-psram.c:96] psramInit(): PSRAM enabled
[I] AudioTypes.h : 127 - in: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 - out: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 -  sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] I2SConfigESP32.h : 80 - rx/tx mode: RXTX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 88200
[I] I2SConfigESP32.h : 84 - bits per sample: 32
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 95 - fixed_mclk: 22579200
[I] I2SConfigESP32.h : 97 - buffer_count:30
[I] I2SConfigESP32.h : 98 - buffer_size:384
[I] I2SConfigESP32.h : 107 - pin_data: 23
[   544][I][BluetoothA2DPCommon.cpp:246] log_free_heap(): [BT_AV] Available Heap: 4216495
[   544][I][BluetoothA2DPSink.cpp:102] start(): [BT_AV] Device name will be set to 'a2dp-spdif'
[   551][I][BluetoothA2DPCommon.cpp:166] get_last_connection(): [BT_AV] nvs_blob does not exist
[   559][I][BluetoothA2DPSink.cpp:165] init_i2s(): [BT_AV] init_i2s
[   565][I][BluetoothA2DPOutput.cpp:199] begin(): [BT_AV] begin
[I] AudioTypes.h : 127 - out: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 -  sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] I2SConfigESP32.h : 80 - rx/tx mode: RXTX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 88200
[I] I2SConfigESP32.h : 84 - bits per sample: 32
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 95 - fixed_mclk: 22579200
[I] I2SConfigESP32.h : 97 - buffer_count:30
[I] I2SConfigESP32.h : 98 - buffer_size:384
[I] I2SConfigESP32.h : 107 - pin_data: 23
[   739][I][BluetoothA2DPSink.cpp:191] init_bluetooth(): [BT_AV] controller initialized
[   740][E][BluetoothA2DPSink.cpp:197] init_bluetooth(): [BT_AV] Failed to initialize bluedroid
[   745][E][BluetoothA2DPSink.cpp:330] app_task_start_up(): [BT_API] app_task_start_up failed
[   753][I][BluetoothA2DPSink.cpp:160] start(): [BT_AV] IDF Version 4.4
[   759][I][BluetoothA2DPCommon.cpp:246] log_free_heap(): [BT_AV] Available Heap: 4194247

**Reply by pschatzmann** [reply]
Strange: I have never seen esp_bluedroid_init() fail!
What ESP32 Core Version do you use ?
Try to deactivate PSRAM...

**Reply by muhammadazmi26** [reply]
ESP32 wrover dev with built-in camera as pictured above. now the camera is removed.

at the bottom of the board it says esp32 wrover dev v1.6

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/41564644/7bb36d06-c66c-4722-9745-0c541fa6045c)

========== This is the serial monitor result after PSRAM is disabled ===========

ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13260
load:0x40080400,len:3028
entry 0x400805e4
[I] AudioTypes.h : 127 - in: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 - out: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 -  sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] I2SConfigESP32.h : 80 - rx/tx mode: RXTX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 88200
[I] I2SConfigESP32.h : 84 - bits per sample: 32
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 95 - fixed_mclk: 22579200
[I] I2SConfigESP32.h : 97 - buffer_count:30
[I] I2SConfigESP32.h : 98 - buffer_size:384
[I] I2SConfigESP32.h : 107 - pin_data: 23
[    96][I][BluetoothA2DPCommon.cpp:246] log_free_heap(): [BT_AV] Available Heap: 24504
[    96][I][BluetoothA2DPSink.cpp:102] start(): [BT_AV] Device name will be set to 'a2dp-spdif'
[   102][I][BluetoothA2DPCommon.cpp:166] get_last_connection(): [BT_AV] nvs_blob does not exist
[   111][I][BluetoothA2DPSink.cpp:165] init_i2s(): [BT_AV] init_i2s
[   117][I][BluetoothA2DPOutput.cpp:199] begin(): [BT_AV] begin
[I] AudioTypes.h : 127 - out: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 -  sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] I2SConfigESP32.h : 80 - rx/tx mode: RXTX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 88200
[I] I2SConfigESP32.h : 84 - bits per sample: 32
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 95 - fixed_mclk: 22579200
[I] I2SConfigESP32.h : 97 - buffer_count:30
[I] I2SConfigESP32.h : 98 - buffer_size:384
[I] I2SConfigESP32.h : 107 - pin_data: 23
[   295][I][BluetoothA2DPSink.cpp:191] init_bluetooth(): [BT_AV] controller initialized
[   296][E][BluetoothA2DPSink.cpp:197] init_bluetooth(): [BT_AV] Failed to initialize bluedroid
[   300][E][BluetoothA2DPSink.cpp:330] app_task_start_up(): [BT_API] app_task_start_up failed
[   308][I][BluetoothA2DPSink.cpp:160] start(): [BT_AV] IDF Version 4.4
[   315][I][BluetoothA2DPCommon.cpp:246] log_free_heap(): [BT_AV] Available Heap: 2252

**Comment by pschatzmann** [comment]
What is your ESP32 Core Version ?

In Arduino go to Tools -> Boards -> Board Manager and type "ESP32"

You will the see e.g. esp32 by Espressif version 2.0.11

The actual version is 2.0.15!

**Reply by muhammadazmi26** [reply]
![image](https://github.com/pschatzmann/arduino-audio-tools/assets/41564644/660171aa-e9f0-49b3-93b1-30faaebb2e1a)

**Reply by pschatzmann** [reply]
So, please upgrade to a reasonable version!

**Reply by muhammadazmi26** [reply]
I have updated version 2.0.15 !
I have re-download [ESP32-A2DP](https://github.com/pschatzmann/ESP32-A2DP) and [arduino-audio-tools](https://github.com/pschatzmann/arduino-audio-tools)

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/41564644/79ee9a58-56e4-4a93-87ac-b0a7900b8223)


ets Jun  8 2016 00:22:57

rst:0x3 (SW_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13232
load:0x40080400,len:3028
entry 0x400805e4
[I] AudioTypes.h : 127 - in: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 - out: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 -  sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] I2SConfigESP32.h : 80 - rx/tx mode: RXTX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 88200
[I] I2SConfigESP32.h : 84 - bits per sample: 32
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 95 - fixed_mclk: 22579200
[I] I2SConfigESP32.h : 97 - buffer_count:30
[I] I2SConfigESP32.h : 98 - buffer_size:384
[I] I2SConfigESP32.h : 107 - pin_data: 23
[   113][I][BluetoothA2DPCommon.cpp:246] log_free_heap(): [BT_AV] Available Heap: 23204
[   121][I][BluetoothA2DPSink.cpp:102] start(): [BT_AV] Device name will be set to 'a2dp-spdif'
[   130][I][BluetoothA2DPCommon.cpp:166] get_last_connection(): [BT_AV] nvs_blob does not exist
[   139][I][BluetoothA2DPSink.cpp:165] init_i2s(): [BT_AV] init_i2s
[   145][I][BluetoothA2DPOutput.cpp:199] begin(): [BT_AV] begin
[I] AudioTypes.h : 127 - out: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 -  sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] I2SConfigESP32.h : 80 - rx/tx mode: RXTX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 88200
[I] I2SConfigESP32.h : 84 - bits per sample: 32
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 95 - fixed_mclk: 22579200
[I] I2SConfigESP32.h : 97 - buffer_count:30
[I] I2SConfigESP32.h : 98 - buffer_size:384
[I] I2SConfigESP32.h : 107 - pin_data: 23
E (201) phy_init: failed to allocate memory for RF calibration data

abort() was called at PC 0x4017f2a1 on core 1


Backtrace: 0x400839ad:0x3ffcbfe0 0x40095001:0x3ffcc000 0x4009a781:0x3ffcc020 0x4017f2a1:0x3ffcc0a0 0x4017f389:0x3ffcc0d0 0x400ea47d:0x3ffcc0f0 0x400d8fcc:0x3ffcc110 0x400d7bdf:0x3ffcc160 0x400d69a7:0x3ffcc190 0x400d4075:0x3ffcc1f0 0x400da402:0x3ffcc260




ELF file SHA256: fbb445813cfbbd72

Rebooting...
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13232
load:0x40080400,len:3028
entry 0x400805e4

========= this keeps repeating ===========

**Comment by muhammadazmi26** [comment]
Then I tried activating PSRAM, here are the results

==================================================================

ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13232
load:0x40080400,len:3028
entry 0x400805e4
[   459][I][esp32-hal-psram.c:96] psramInit(): PSRAM enabled
[I] AudioTypes.h : 127 - in: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 - out: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 -  sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] I2SConfigESP32.h : 80 - rx/tx mode: RXTX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 88200
[I] I2SConfigESP32.h : 84 - bits per sample: 32
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 95 - fixed_mclk: 22579200
[I] I2SConfigESP32.h : 97 - buffer_count:30
[I] I2SConfigESP32.h : 98 - buffer_size:384
[I] I2SConfigESP32.h : 107 - pin_data: 23
[   553][I][BluetoothA2DPCommon.cpp:246] log_free_heap(): [BT_AV] Available Heap: 4215151
[   563][I][BluetoothA2DPSink.cpp:102] start(): [BT_AV] Device name will be set to 'a2dp-spdif'
[   571][I][BluetoothA2DPCommon.cpp:166] get_last_connection(): [BT_AV] nvs_blob does not exist
[   580][I][BluetoothA2DPSink.cpp:165] init_i2s(): [BT_AV] init_i2s
[   586][I][BluetoothA2DPOutput.cpp:199] begin(): [BT_AV] begin
[I] AudioTypes.h : 127 - out: sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] AudioTypes.h : 127 -  sample_rate: 88200 / channels: 2 / bits_per_sample: 32
[I] I2SConfigESP32.h : 80 - rx/tx mode: RXTX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 88200
[I] I2SConfigESP32.h : 84 - bits per sample: 32
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 95 - fixed_mclk: 22579200
[I] I2SConfigESP32.h : 97 - buffer_count:30
[I] I2SConfigESP32.h : 98 - buffer_size:384
[I] I2SConfigESP32.h : 107 - pin_data: 23
[   731][I][BluetoothA2DPSink.cpp:191] init_bluetooth(): [BT_AV] controller initialized
[   739][E][BluetoothA2DPSink.cpp:197] init_bluetooth(): [BT_AV] Failed to initialize bluedroid
[   748][E][BluetoothA2DPSink.cpp:330] app_task_start_up(): [BT_API] app_task_start_up failed
[   756][I][BluetoothA2DPSink.cpp:160] start(): [BT_AV] IDF Version 4.4
[   763][I][BluetoothA2DPCommon.cpp:246] log_free_heap(): [BT_AV] Available Heap: 4192891

**Comment by pschatzmann** [comment]
You don't have enough RAM: try to reduce the cfg.buffer_count!

**Reply by muhammadazmi26** [reply]
i try to set cfg.buffer_count = 20; it just doesn't work like before. 
Then buffer_count = 8 Doesn't work.

i try reduce buffer_size: 128 Doesn't work 😫

...
...
E (201) phy_init: failed to allocate memory for RF calibration data

abort() was called at PC 0x4017f2a9 on core 1
....
....

**Reply by pschatzmann** [reply]
I tend to think that this is a problem with your hardware.

**Reply by muhammadazmi26** [reply]
Okay
thank you very much for helping
I will try buying a new one later for comparison

**Comment by pschatzmann** [comment]
For me both buffer_count 30 and PSRAM active and not active is working.

But I noticed that reducing the buffer_count gets overwritten by the default value of 30!
I committed a [correction to the AudioTools](https://github.com/pschatzmann/arduino-audio-tools/commit/47c28c4619de55d9125414c34df9f75dae942189) and I suggest that you try again with the corrected version. If you check in the log you should see now the correct values:
09:49:43.791 -> [I] I2SConfigESP32.h : 97 - buffer_count:8
09:49:43.791 -> [I] I2SConfigESP32.h : 98 - buffer_size:384

**Comment by muhammadazmi26** [comment]
Yaa, it works 😁
Thanks 

This is the video when i tried it:
https://github.com/pschatzmann/arduino-audio-tools/assets/41564644/49a36404-c11a-443b-8835-c3b89f1106ae
